precompiles/staking/method_claim_rewards.go

```diff
@@ -0,0 +1,147 @@
+package staking
+
+import (
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+
+	"github.com/zeta-chain/node/cmd/zetacored/config"
+	"github.com/zeta-chain/node/precompiles/bank"
+	precompiletypes "github.com/zeta-chain/node/precompiles/types"
+	fungibletypes "github.com/zeta-chain/node/x/fungible/types"
+)
+
+// claimRewards claims all the rewards for a delegator from a validator.
+// As F1 Cosmos distribution scheme implements an all or nothing withdrawal, the precompile will
+// withdraw all the rewards for the delegator, filter ZRC20 and unlock them to the delegator EVM address.
+func (c *Contract) claimRewards(
+	ctx sdk.Context,
+	evm *vm.EVM,
+	contract *vm.Contract,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	if len(args) != 2 {
+		return nil, &precompiletypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 2,
+		}
+	}
+
+	delegatorAddr, validatorAddr, err := unpackClaimRewardsArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		// This represents the delegator calling directly the precompile.
+		callerIsDelegator = contract.CallerAddress == delegatorAddr
+
+		// This represents the delegator calling the precompile through a contract.
+		originIsDelegator = evm.Origin == delegatorAddr
```

Comment:
I'm not sure I understand this one. If you build a smart contract application to stake, the smart contract would be the delegator, then the smart contract would have the responsibility to distribute stake, etc..
I don't think we should ever use Origin in any case.
Right now since stake function is disable a smart contract can't stake so we consider only EoA will be able to use this function

Reply:
With this approach we ensure that the delegator (the EOA) is somehow involved in claiming the rewards. Either it is the origin of the transaction, or the immediate caller of the claimRewards function. 
I think it's a good solution to avoid possible lose of funds.

If we allow smart contracts to be delegators, how would an user recover its funds and the ZRC20? The dApp would have control over all the funds.

Reply:
>If we allow smart contracts to be delegators, how would an user recover its funds and the ZRC20? The dApp would have control over all the funds.

The DApp is supposed to define the logic for this